Sources/google-cli-swift-generator/main.swift

```diff
@@ -0,0 +1,309 @@
+    s.addLine(indent:10, "sem.signal()")
+    s.addLine(indent:8, "}")
+    s.addLine(indent:8, "_ = sem.wait()")
+    s.addLine(indent:6, "} catch let error {")
+    s.addLine(indent:8, "print (\"Client error: \\(error)\")")
+    s.addLine(indent:6, "}")
+    s.addLine(indent:4, "}")
+    return s
+  }
+}
+
+extension Discovery.Service {
+  func serviceTitle() -> String {
+    return self.name.capitalized()
+  }
+  func serviceName() -> String {
+    return self.name
+  }
+  func scopes() -> [String] {
+    var scopeSet = Set<String>()
+    if let resources = resources {
+      for r in resources {
+        let methods = r.value.methods
+        for m in methods {
+          if let scopes = m.value.scopes {
+            for scope in scopes {
+              scopeSet.insert(scope)
+            }
+          }
+        }
+      }
+    }
+    return scopeSet.sorted()
+  }
+  func generate() -> String {
+    var s = Discovery.License
+    s.addLine()
+    for i in
+      ["Foundation",
+       "Dispatch",
+       "OAuth2",
+       "GoogleAPIRuntime",
+       "Commander"] {
+        s.addLine("import " + i)
+    }
+    s.addLine()
+    s.addLine("let CLIENT_CREDENTIALS = \"" + serviceName() + ".json\"")
+    s.addLine("let TOKEN = \"" + serviceName() + ".json\"")
+    s.addLine()
+    s.addLine("func main() throws {")
+    let scopes = self.scopes()
+    if scopes.count == 1 {
+      s.addLine(indent:2, "let scopes = \(scopes)")
+    } else {
+      s.addLine(indent:2, "let scopes = [")
+      s += "    \"" + scopes.joined(separator:"\",\n    \"") + "\"]\n"
+    }
+    s.addLine()
+    s.addLine(indent:2, "guard let tokenProvider = BrowserTokenProvider(credentials:CLIENT_CREDENTIALS, token:TOKEN) else {")
+    s.addLine(indent:4, "return")
+    s.addLine(indent:2, "}")
+    s.addLine(indent:2, "let \(self.serviceName()) = try \(self.serviceTitle())(tokenProvider:tokenProvider)")
+    s.addLine()
+    s.addLine(indent:2, "let group = Group {")
+    s.addLine(indent:4, "$0.command(\"login\", description:\"Log in with browser-based authentication.\") {")
+    s.addLine(indent:6, "try tokenProvider.signIn(scopes:scopes)")
+    s.addLine(indent:6, "try tokenProvider.saveToken(TOKEN)")
+    s.addLine(indent:4, "}")
+    if let resources = resources {
+      for r in resources.sorted(by:  { $0.key < $1.key }) {
+        let methods = r.value.methods
+        for m in methods.sorted(by:  { $0.key < $1.key }) {
+          let requestSchema = self.schema(name: m.value.RequestTypeName())
+          s += m.value.Invocation(serviceName:self.serviceName(),
+                                  resourceName:r.key,
+                                  resource:r.value,
+                                  methodName:m.key,
+                                  method:m.value,
+                                  requestSchema:requestSchema
+          )
+        }
+      }
+    }
+    s.addLine(indent:2, "}")
```

Comment:
So it looks like you're trying to construct these all with string concatenation.  In the nodejs client, we're using a templating engine ([nunjucks](https://mozilla.github.io/nunjucks/) in my case) and I'm pretty happy with it - especially when you start getting into the weird corners. 

Another route I've considered is using the TypeScript AST APIs to generate this stuff, but initial experiments were MUCH harder to understand.  Anyhow - I'm not opposed to the approach you're taking, I just think it will be limiting as you start to support more features.

Reply:
I started gRPC-Swift with a templating engine and found that it came with a lot of bloat and didn't add much value over smart print statements. The tool that I've seen for generating Swift from ASTs (https://nshipster.com/swiftsyntax/) is interesting but still are pretty verbose and obscure.